### PR TITLE
M2m topics redirects

### PR DESCRIPTION
**Pull request by akatsoulas**

### CODE REVIEW

File: kitsune/products/models.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from tabnanny import verbose
```
[the remark this answers is not in the document]
Comment (akatsoulas) [reply kitsune/products/models.py:1]: sorry for missing that

File: kitsune/products/views.py [hunk 2]
```diff
@@ -76,10 +75,30 @@ def product_landing(request, slug):
...
+    if topic_slug:
+        try:
+            old_topic_slug = TopicSlugHistory.objects.get(slug=topic_slug)
+            redirect_params = {
+                "topic_slug": old_topic_slug.topic.slug,
+            }
+
+            if product_slug:
+                redirect_params["product_slug"] = product_slug
+            if subtopic_slug:
```
[the remark this answers is not in the document]
Comment (akatsoulas) [reply kitsune/products/views.py:95]: Great question! We might need it in the second iteration but we don't for the first one which will include the redirects for the navigation menus. I figured that we can add code and complexity when/if we need it.

File: kitsune/products/models.py [hunk 3]
```diff
@@ -169,6 +172,24 @@ def get_absolute_url(self):
             )
 
 
+class ProductTopic(ModelBase):
+    """Through model for Product and Topic to add additional metadata."""
+
+    product = models.ForeignKey(Product, on_delete=models.CASCADE)
+    topic = models.ForeignKey(Topic, on_delete=models.CASCADE)
+    created = models.DateTimeField(auto_now_add=True)
```
[the remark this answers is not in the document]
Comment (akatsoulas) [reply kitsune/products/models.py:180]: Nice catch!

**Review by escattone — COMMENTED** [review]
The only thing I'm not sold on is the `TopicSlugHistory` table.

File: kitsune/products/models.py [hunk 4]
```diff
@@ -169,6 +170,29 @@ def get_absolute_url(self):
...
+
+
+class TopicSlugHistory(ModelBase):
+    topic = models.ForeignKey(Topic, on_delete=models.CASCADE, related_name="slug_history")
+    slug = models.CharField(max_length=255)
+    created = models.DateTimeField(auto_now_add=True)
+
+    class Meta(object):
+        ordering = ["-created"]
+        verbose_name_plural = "Topic slug history"
```
[the remark this answers is not in the document]
Comment (akatsoulas) [reply kitsune/products/models.py:193]: * Having just a ` from_slug` and `to_slug` will not work currently. We have many topics with the same slug and what differentiates them currently is the FK to the product. 
* In the first iteration will need to align and normalize all the slugs for each product. For example `install-and-update-ios` and `install-and-update-firefox` will become `install-and-update`. In this step we will have 2 topics with the same slug. One topic will have one FK to Firefox and the other to Firefox for iOS.
* In the next iteration, we will start introducing the new topics. This iteration won't have multiple topics with the same slug but just one topic for each slug. In this case all the Topic FKs in the SlugHistory entries will be updated to point to the single Topic.
* We can add either unique=True or make this a SlugField for sure.
* The fact that we might change redirects in the future makes me think that probably we don't need to issue a permanent redirect yet.

File: kitsune/products/models.py [hunk 5]
```diff
@@ -169,6 +170,29 @@ def get_absolute_url(self):
...
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=["product", "topic"], name="unique_product_topic")
+        ]
+
+
+class TopicSlugHistory(ModelBase):
+    topic = models.ForeignKey(Topic, on_delete=models.CASCADE, related_name="slug_history")
+    slug = models.SlugField(max_length=255, unique=True)
```
Comment (escattone) [inline kitsune/products/models.py:188]: Do we need to add `allow_unicode=True`? Hm, it appears that our topic slugs are never translated, so probably not.

**Review by escattone — APPROVED** [review]
r+wc

**Pull request merged**